lib/DXIL/DxilShaderModel.cpp

```diff
@@ -76,7 +76,8 @@ bool ShaderModel::IsValidForDxil() const {
 
 bool ShaderModel::IsValidForModule() const {
   // Ray tracing shader model should only be used on functions in a lib
-  return IsValid() && !IsRay();
+  return IsValid() &&
+         !(m_Kind >= Kind::RayGeneration && m_Kind <= Kind::Callable);
```

Comment:
Technically, this exposed a hole.  Node is not a valid kind for module either, so we should update the expression to reflect that, and update the comment, since it's not just raytracing.

Reply:
Actually, there's more.  Notice how `IsValid()` is asserting that it's one of the supported kinds?  `IsValidForModule()` calls `IsValid()`.  If it was ray or node, it would assert (on debug) in `IsValid()`.

It looks like the only way you could have a kind outside the valid or the explicit `Kind::Invalid`, is to manually construct the ShaderModel, but that constructor is private.  So, I think the assert is sufficient, and there's no need for additional checks in this method beyond the current IsValid().

Then, if `IsValidForModule()` is identical to `IsValid()`, why have it?  There's no reason to keep it.  The only use of this function is in an assert in `DxilModule::SetShaderModel`.  Since the prior assert already calls `IsValidForDxil()`, which calls `IsValid()` as well, there's no need for the assert that calls `IsValidForModule()`.

In short, we should be able to remove this method and the assert where it's called.